### DISCUSSION TITLE
Getting different trace Id when using @ControllerAdvice in Java with Opentelemetry

### DISCUSSION THREAD

**Discussion by keerthanacv** [open-telemetry/opentelemetry-java-instrumentation#12447] [Q&A] [answered]
I'm using @ControllerAdvice to handle all the exceptions in java, but I'm receiving a different trace Id, when an exception is thrown

@ControllerAdvice
public class RestExceptionHandler extends ResponseEntityExceptionHandler {

    @Autowired
    private Tracer tracer;

    @Autowired
    OpentelemetryService opentelemetryService;

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<Object> handleResourceNotFoundException(ResourceNotFoundException resourceNotFoundException,
                                                                  HttpServletRequest request) {

        Span exceptionHandlerSpan = Span.current();
        //try (Scope scope = exceptionHandlerSpan.makeCurrent()) {
            ErrorDetail errorDetail = new ErrorDetail(HttpStatus.NO_CONTENT, "NO_RESULTS", resourceNotFoundException.getErrCode(),
                    "No Data Found", resourceNotFoundException.getMessage());

            CommonResponse response = new CommonResponse(true, PortalConstants.FAILURE,
                    ObjectUtils.isNotEmpty(exceptionHandlerSpan.getSpanContext().getTraceId())
                            ? exceptionHandlerSpan.getSpanContext().getTraceId()
                            : EMPTY_STRING
                    , Arrays.asList(errorDetail), new ArrayList<>());
            exceptionHandlerSpan.end();
            return new ResponseEntity<>(response, null, HttpStatus.OK);
     //   }
    }
    
    
    Controller class :
    
    @RequestMapping(value = "/v2/getdetails", method = RequestMethod.POST)
    public ResponseEntity<EventChartsDetailsResponse> getEventTypesChartDetailsV2(
            @RequestHeader Map<String, Object> headers,
            @Valid @RequestBody ChartViewRequest chartViewRequest) throws IOException {

        Span span = opentelemetryService.controllerSpan("PortalController", tracer,
                (String) headers.get("request-id"), (String) headers.get("trace-id"));
        try (Scope scope = span.makeCurrent()) {
            LOGGER.info("In PortalController :: getEventTypesChartDetailsV2");
            EventChartsDetailsResponse response = null;
            if (Objects.nonNull(chartViewRequest)) {
                try {
               //logic
                    } 
                } catch (ResourceNotFoundException e) {
                    ResourceNotFoundException error =
                            new ResourceNotFoundException("No Data Found !");
                    error.setErrCode(ErrorConstants.NOT_FOUNR);
                    LOGGER.error(" error);
                    throw error;
                } 
            }
            return new ResponseEntity<>(response, HttpStatus.OK);
        } catch (Exception e) {
            throw e;
        } finally {
            opentelemetryService.endControllerSpan(span);
        }
        
        How do I fix it?

**Comment by laurit** [comment]
Please provide minimal application that reproduces your issue along with any instructions necessary.

**Reply by keerthanacv** [reply]
[controllerAdvice.zip](https://github.com/user-attachments/files/17395089/controllerAdvice.zip)


Trying hitting the endpoint curl

curl --location 'http://localhost:8080/getString' \
--header 'Content-Type: application/json' \
--data '{
    "name":""
}'

And observe the logs 
2024-10-16 17:17:56.388 INFO   Correlation-ID : c31c4251b7df8cebf3ec89c94b221b29 Span-Id : 8df5eadc47946bd4 Request-Id:  Trace-ID :   Class : PortalController  logger failure
2024-10-16 17:17:56.392 INFO   Correlation-ID : c31c4251b7df8cebf3ec89c94b221b29 Span-Id : 8df5eadc47946bd4 Request-Id:  Trace-ID :   Class : PortalController  'PortalControllernull' : c31c4251b7df8cebf3ec89c94b221b29 8df5eadc47946bd4 INTERNAL [tracer: CommCentral Application:] {}
2024-10-16 17:17:56.478 INFO   Correlation-ID : 587d05926858986ad5246ea605a69649 Span-Id : 5a919cb3817e1c4e Request-Id:  Trace-ID :   Class : RestExceptionHandler  'RestExceptionHandler' : 587d05926858986ad5

I need same Id c31c4251b7df8cebf3ec89c94b221b29 in rest exceptionhandler class as well

**Reply by laurit** [reply, marked as answer]
The reason you get separate traces is that by the time the exception handler is run your span in `PortalController` is already completed. Since there is no active trace creating a new span in the exception handler will also start a new trace. To overcome this you can either somehow pass your span from the `PortalController` to the exception handler so you could use it as the parent span, which is probably not a good idea, or you could create a span that surrounds the whole request and would be the parent for both the span you create in `PortalController` and exception handler. My advice to you would be to use the spring boot starter or the java agent https://opentelemetry.io/docs/zero-code/java/ They'll create a span for the whole request which should solve your issue.
